### PR TITLE
audit: add rules on field order

### PR DESCRIPTION
**Pull request by xu-cheng**
Per https://github.com/Homebrew/homebrew/pull/40469#issue-85876583

cc @Homebrew/owners @bfontaine @DomT4

### CODE REVIEW

File: Library/Homebrew/cmd/audit.rb [hunk 1]
```diff
@@ -56,6 +56,7 @@ def audit
 class FormulaText
   def initialize path
     @text = path.open("rb", &:read)
+    @lines = @text.split("\n")
```
Comment (mikemcquaid) [inline Library/Homebrew/cmd/audit.rb]: `@text.lines`

**Comment by mikemcquaid** [comment]
Makes sense to me but I suspect other @Homebrew/owners may have opinions

File: Library/Homebrew/cmd/audit.rb (hunk 1 above)
Comment (xu-cheng) [inline Library/Homebrew/cmd/audit.rb]: `lines` will retain `\n` so I use split.

Comment (mikemcquaid) [inline Library/Homebrew/cmd/audit.rb]: Mmm, still feels a bit weird. Does the presence of `\n` matter below?

Comment (xu-cheng) [inline Library/Homebrew/cmd/audit.rb]: > Does the presence of \n matter below?

En, seems not matter. So I have changed it into `lines`.

**Comment by DomT4** [comment]
Do we want to make a decision/enforcement about `version` line placement as well? We typically allow it either below the `url` or below the `sha`.

**Comment by xu-cheng** [comment]
I rewrite the code to make it more scalable and add more field into the rules.

File: Library/Homebrew/cmd/audit.rb [hunk 2]
```diff
@@ -141,6 +147,36 @@ def audit_file
...
+      [/^  desc ["'][\S\ ]+["']/,          "desc"          ],
+      [/^  homepage ["'][\S\ ]+["']/,      "homepage"      ],
+      [/^  url ["'][\S\ ]+["']/,           "url"           ],
+      [/^  mirror ["'][\S\ ]+["']/,        "mirror"        ],
+      [/^  version ["'][\S\ ]+["']/,       "version"       ],
+      [/^  (sha1|sha256) ["'][\S\ ]+["']/, "checksum"      ],
+      [/^  head ["'][\S\ ]+["']/,          "head"          ],
+      [/^  stable do/,                     "stable block"  ],
+      [/^  devel do/,                      "devel block"   ],
+      [/^  head do/,                       "head block"    ],
```
Comment (mikemcquaid) [inline Library/Homebrew/cmd/audit.rb]: Maybe add the bottle block in here?

**Comment by mikemcquaid** [comment]
Looks great to me.

File: Library/Homebrew/cmd/audit.rb (hunk 2 above)
Comment (xu-cheng) [inline Library/Homebrew/cmd/audit.rb]: Added.

**Comment by mikemcquaid** [comment]
:+1:

**Comment by tdsmith** [comment]
This seems to be breaking on Mountain Lion (i.e. Ruby 1.8): http://bot.brew.sh/job/Homebrew%20Science%20Pull%20Requests/2188/version=mountain_lion/testReport/junit/brew-test-bot/mountain_lion/audit_Homebrew_science_boost_compute___strict/

**Comment by DomT4** [comment]
^ Seeing the same thing at X11 as well: http://bot.brew.sh/job/Homebrew%20X11%20Pull%20Requests/148/version=mountain_lion/console

**Comment by xu-cheng** [comment]
Should be fixed in fc20f215f3bf104f7f957bb9159d8b70cc890688

**Comment by mikemcquaid** [comment]
We should probably add more test runs of various commands e.g. `brew audit` in our test suite and/or `brew test-bot` at some point.

**Pull request closed without merging**